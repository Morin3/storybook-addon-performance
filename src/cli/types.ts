import { ResultMap } from '../types';
import { performCalculations } from './utils';

export type Results = {
  [key in keyof ResultMap]: number[];
};

export type ResultsByGroupId = { [groupId: string]: Results };
export interface Calculation {
  key: string;
  numberOfSamples: number;
  samples: number[];
  minValue: number;
  maxValue: number;
  meanValue: number;
  medianValue: number;
}

export type CalculationsByGroupId = { [groupId: string]: Calculation[] };
export type CalculationsByDirectory = { [name: string]: CalculationsByGroupId };
