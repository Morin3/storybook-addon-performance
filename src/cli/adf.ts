import * as fs from 'fs';

import { buildTable, buildAdf, buildTableRows } from './util/adf';
import { CalculationsByGroupId } from './types';
import { debug, stdout } from './util/print';
import { writeFile } from './util/write';

const adf = () => {
  if (process.argv.length <= 2) {
    return stdout('ðŸ’” Oh no! Please provide a path to the input file.');
  }

  /**
   * Get arguments passed to find the file
   * with the formatted output.
   */
  const [inputPath] = process.argv.slice(2);
  const [inputFileName] = inputPath.split('.');

  const data = fs.readFileSync(`${inputFileName}.json`, 'utf-8');

  try {
    /**
     * Build one table + heading for each test group ("Server", "Client", "Interactions"),
     * containing measurements for lite mode + its baseline, and
     * the difference as a percent.
     */
    const calculationsByGroupId = JSON.parse(data) as CalculationsByGroupId;
    const outputTables = Object.entries(calculationsByGroupId)
      .map(([key, results]) => buildTable(key, results.map(buildTableRows)))
      .flatMap((table) => table);

    const adf = buildAdf(outputTables);

    const outputPath = `${inputFileName}-adf.json`;
    const content = JSON.stringify(adf);

    writeFile(outputPath, content, `Generated ADF is saved to ${outputPath}!`);
  } catch (e) {
    debug(
      `ðŸ’” Problem parsing a file in '${inputPath}' - was this created by the storybook-addon-performance? \n`,
      e,
    );
  }
};

adf();
