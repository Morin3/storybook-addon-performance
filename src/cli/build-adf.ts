import { Calculation } from './types';

interface Attributes {
  title?: string;
  text?: string;
  color?: string;
  style?: string;
  background?: string;
  isNumberColumnEnabled?: boolean;
  layout?: string;
}

export interface Content {
  type: string;
  attrs?: Attributes;
  text?: string;
  content?: Content[];
  marks?: {
    type: string;
  }[];
}

export const buildAdf = (content: Content[]) => ({
  version: 1,
  type: 'doc',
  content,
});

export const buildTable = (heading: string, tableRows: Content[]) => [
  {
    type: 'heading',
    attrs: {
      level: 2,
    },
    content: [
      {
        type: 'text',
        text: heading,
      },
    ],
  },
  {
    type: 'table',
    attrs: {
      isNumberColumnEnabled: false,
      layout: 'default',
    },
    content: [
      {
        type: 'tableRow',
        content: [
          {
            type: 'tableHeader',
            content: [
              {
                type: 'paragraph',
                content: [],
              },
            ],
          },
          {
            type: 'tableHeader',
            content: [
              {
                type: 'paragraph',
                content: [
                  {
                    type: 'text',
                    text: 'Baseline',
                    marks: [
                      {
                        type: 'strong',
                      },
                    ],
                  },
                ],
              },
              {
                type: 'paragraph',
                content: [
                  {
                    type: 'text',
                    text: 'This should test the component before any changes.',
                    marks: [
                      {
                        type: 'subsup',
                        attrs: {
                          type: 'sup',
                        },
                      },
                    ],
                  },
                ],
              },
            ],
          },
          {
            type: 'tableHeader',
            content: [
              {
                type: 'paragraph',
                content: [
                  {
                    type: 'text',
                    text: 'Lite mode',
                    marks: [
                      {
                        type: 'strong',
                      },
                    ],
                  },
                ],
              },
              {
                type: 'paragraph',
                content: [
                  {
                    type: 'text',
                    text: 'This should test the fully converted component.',
                    marks: [
                      {
                        type: 'subsup',
                        attrs: {
                          type: 'sup',
                        },
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      ...tableRows,
    ],
  },
];

export const buildNameCell = (key: string): Content => {
  return {
    type: 'tableCell',
    attrs: {
      background: '#f4f5f7',
    },
    content: [
      {
        type: 'paragraph',
        content: [
          {
            type: 'text',
            text: key,
            marks: [
              {
                type: 'strong',
              },
            ],
          },
        ],
      },
    ],
  };
};

export const buildResultCell = (
  { medianValue, numberOfSamples, samples }: Omit<Calculation, 'key'>,
  diff?: number,
): Content => {
  return {
    type: 'tableCell',
    content: [
      {
        type: 'paragraph',
        content: [
          {
            type: 'text',
            text: medianValue.toFixed(2),
          },
          ...buildDiffStatus(diff),
        ],
      },
      {
        type: 'nestedExpand',
        attrs: {
          title: `Raw runs (${numberOfSamples} times)`,
        },
        content: [
          {
            type: 'paragraph',
            content: [
              {
                type: 'text',
                text: samples.map((num: number) => num.toFixed(2)).join('\t'),
              },
            ],
          },
        ],
      },
    ],
  };
};

const buildDiffStatus = (diff?: number) => {
  if (!diff) {
    return [];
  }

  return [
    {
      type: 'text',
      text: ' ',
    },
    {
      type: 'status',
      attrs: {
        text: String(diff.toFixed(2)) + '%',
        color: 'neutral',
        style: 'bold',
      },
    },
    {
      type: 'text',
      text: ' ',
    },
  ];
};
