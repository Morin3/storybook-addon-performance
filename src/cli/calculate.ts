import * as path from 'path';

import {
  CalculationsByDirectory,
  CalculationsByGroupId,
  ProcessDescription,
  ResultsByGroupId,
} from './types';
import { performAllCalculations } from './util/calculate';
import { writeFile } from './util/write';

const calculate = (resultsByDirectory: (ResultsByGroupId & { name: string })[]) => {
  /**
   * Calculate the max, min, median, and mean values
   * of the previously formatted outputs, and group the results
   * per directory.
   */
  const calculationsByDirectory = resultsByDirectory.reduce(
    (calculationsByDirectory, { name, ...resultsByGroupId }) => ({
      ...calculationsByDirectory,
      [path.basename(name)]: Object.entries(resultsByGroupId).reduce(
        performAllCalculations,
        {} as CalculationsByGroupId,
      ),
    }),
    {} as CalculationsByDirectory,
  );

  /**
   * Write the results into files â€“
   * one file summarises one input directory.
   */
  Object.entries(calculationsByDirectory).forEach(([directoryName, output]) => {
    writeFile(ProcessDescription.Calculate, directoryName, JSON.stringify(output));
  });

  /**
   * Return the results to be compared against each other.
   */
  return calculationsByDirectory;
};

export default calculate;
