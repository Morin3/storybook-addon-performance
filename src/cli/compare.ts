import * as fs from 'fs';
import { Calculation, CalculationsByGroupId } from './types';
import { calculateDifference } from './util/compare';
import { debug, stdout } from './util/print';
import { writeFile } from './util/write';

const compare = () => {
  if (process.argv.length <= 2) {
    return stdout('ðŸ’” Oh no! Please provide paths to the input files.');
  }

  /**
   * Get arguments passed to figure out which
   * input is the baseline.
   */
  const input = process.argv.slice(2);
  const categories = { '-l': 'Lite mode', '-b': 'Baseline' };
  const flags = Object.keys(categories);

  const inputPaths = input
    .map((arg, i) => (flags.includes(arg) ? { [arg]: input[i + 1] } : {}))
    .reduce((acc, val) => ({ ...acc, ...val }));

  const baselineString = fs.readFileSync(inputPaths['-b'], 'utf-8');
  const liteString = fs.readFileSync(inputPaths['-l'], 'utf-8');

  try {
    const baseline = JSON.parse(baselineString) as CalculationsByGroupId;
    const lite = JSON.parse(liteString) as CalculationsByGroupId;

    /**
     * Calculate the difference between the median value of each input,
     * and aggregate the output per test group ("Server", "Client", "Interactions").
     */
    const allDiffs = Object.entries(lite)
      .map(
        ([groupId, lite]) =>
          [groupId, lite, baseline[groupId]] as [string, Calculation[], Calculation[]],
      )
      .map(([groupId, lite, baseline]) => ({
        [groupId]: lite.map(calculateDifference(baseline)),
      }))
      .reduce((acc, val) => ({ ...acc, ...val }));

    const [baselinePath] = inputPaths['-b'].split('.');
    const [litePath] = inputPaths['-l'].split('.');

    const outputPath = `${litePath}-vs-${baselinePath}.json`;
    const content = JSON.stringify(allDiffs);

    writeFile(outputPath, content, `Difference is calculated and saved to ${outputPath}!`);
  } catch (e) {
    debug(
      `ðŸ’” Problem parsing a file in '${baselineString} and ${liteString}' - were these created by the storybook-addon-performance? \n`,
      e,
    );
  }
};

compare();
