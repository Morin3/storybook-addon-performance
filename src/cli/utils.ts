/* eslint-disable no-console */
import type { TaskGroupResult } from '../types';
import { Results } from './types';

/**
 * Flattens out the result to a form we can consume
 */
export const processResult = (result: TaskGroupResult) => {
  return Object.keys(result.map).map((key) => {
    // @ts-ignore
    const { taskName, averageMs, value } = result.map[key];
    return {
      taskName,
      value: averageMs || Number(value),
    };
  });
};

/**
 * Turns the row output into a csv
 */
export const rowify = (data: Results, numResults: number) => {
  console.info(
    `type,${Array.from({ length: numResults })
      .map((_, i) => `#${i + 1}`)
      .join(',')},min,max,ave`,
  );
  Object.entries(data).forEach(([key, values]) => {
    console.info(
      `${key}, ${values.join(',')},${Math.min(...values)}, ${Math.max(...values)}, ${
        values.reduce((acc, curr) => acc + curr, 0) / 5
      }`,
    );
  });
};

export const usage = () =>
  console.warn(`Usage:
storybook-addon-performance <directory> [...<directory>] 

Example
storybook-addon-performance results-directory > outputfile.csv
# OR
storybook-addon-performance ABTestDirectory OtherDirectory > outputfile.csv
`);
