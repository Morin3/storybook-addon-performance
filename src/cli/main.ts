import * as fs from 'fs';
import * as path from 'path';

import type { TaskGroupResult } from '../types';

import calculate from './calculate';
import compare from './compare';
import generateAdf from './adf';

import type { ResultsByGroupId, Results } from './types';
import { convertToTaskValueMap, combineTaskResultsByGroupId } from './util/calculate';
import { debug, usage } from './util/print';

const main = (...args: string[]) => {
  /**
   * Get storybook-addon-performance result files
   * in the specified directories.
   */
  const cliArgs = args.length ? args : process.argv;
  if (cliArgs.length <= 2) {
    return usage();
  }

  const input = cliArgs.slice(2);
  const categories = { '-l': 'Lite mode', '-b': 'Baseline' };
  const flags = Object.keys(categories);

  const inputPaths = input
    .map((arg, i) => (flags.includes(arg) ? { [arg]: input[i + 1] } : {}))
    .reduce((acc, val) => ({ ...acc, ...val }));

  if (Object.entries(inputPaths).length < 2) {
    return usage();
  }

  const dirPaths = [inputPaths['-b'], inputPaths['-l']];
  const resultsByDirectory = dirPaths.map((pathName) => {
    const files = fs.readdirSync(pathName, 'utf-8');
    if (!files) {
      return debug(
        `ðŸ’” Directory '${pathName}' is empty - ` +
          'did you specify a directory with storybook-addon-performance output files?',
      );
    }

    try {
      /**
       *  Format and group the inputs per directory.
       */
      const resultsByGroupId = files
        .map((dataFile) => {
          const json = fs.readFileSync(path.join(pathName, dataFile));
          return JSON.parse(json as any);
        })
        .map(({ results }) => results as TaskGroupResult[])
        .flatMap((taskGroupResults) => taskGroupResults)
        .map(({ groupId, map }) => [groupId, convertToTaskValueMap(map)] as [string, Results])
        .reduce(combineTaskResultsByGroupId, {} as ResultsByGroupId);

      return { name: pathName, ...resultsByGroupId };
    } catch (e) {
      return debug(
        `ðŸ’” Problem parsing a file in '${pathName}' - ` +
          'was this created by the storybook-addon-performance? \n',
        e,
      );
    }
  }) as (ResultsByGroupId & { name: string })[];

  const calculationsByDirectory = calculate(resultsByDirectory);

  const [baseline, lite] = Object.values(calculationsByDirectory);
  const calculationsWithDiff = compare(baseline, lite);

  generateAdf(calculationsWithDiff);
};

export default main;
