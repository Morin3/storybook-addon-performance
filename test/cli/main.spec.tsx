import cli from '../../src/cli/main';

describe('cli', () => {
  afterAll(jest.restoreAllMocks);
  afterEach(jest.resetAllMocks);
  const infoMock = jest.fn();
  const warnMock = jest.fn();
  //@ts-ignore
  global.console = { log: infoMock, warn: warnMock };
  it('should show message if not provided correct args', () => {
    cli();
    expect(warnMock).toHaveBeenCalledWith(`Usage:
sb-perf <directory> [...<directory>]

Example
sb-perf results-directory > output-file.csv
# OR
sb-perf ABTestDirectory OtherDirectory > output-file.csv
`);
  });

  it('should match snapshot without interactions test results', () => {
    cli('node', 'cli', __dirname + '/fixtures/icon');
    expect(infoMock).toHaveBeenCalled();
    expect(infoMock.mock.calls).toMatchSnapshot();
    expect(warnMock.mock.calls).toMatchSnapshot();
  });

  it('should match snapshot with interactions test results', () => {
    cli('node', 'cli', __dirname + '/fixtures/menu');
    expect(infoMock).toHaveBeenCalled();
    expect(infoMock.mock.calls).toMatchSnapshot();
    expect(warnMock.mock.calls).toMatchSnapshot();
  });
});
