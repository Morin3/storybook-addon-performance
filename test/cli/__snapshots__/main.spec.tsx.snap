// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should match snapshot 1`] = `
Array [
  Array [
    "fixtures",
  ],
  Array [
    "Serverside",
  ],
  Array [
    "type,#1,min,max,mean,median",
  ],
  Array [
    "Render to string,11.901000000761996,11.901000000761996,11.901000000761996,11.901000000761996,11.901000000761996",
  ],
  Array [
    "Render to static markup (cannot be hydrated),10.560999999870546,10.560999999870546,10.560999999870546,10.560999999870546,10.560999999870546",
  ],
  Array [
    "String output size,184.64,184.64,184.64,184.64,184.64",
  ],
  Array [
    "String output size (gzip),41.16,41.16,41.16,41.16,41.16",
  ],
  Array [
    "Static markup output size,184.64,184.64,184.64,184.64,184.64",
  ],
  Array [
    "Static markup output size (gzip),41.16,41.16,41.16,41.16,41.16",
  ],
  Array [
    "Clientside",
  ],
  Array [
    "type,#1,min,max,mean,median",
  ],
  Array [
    "Initial render,23.0890000000727,23.0890000000727,23.0890000000727,23.0890000000727,23.0890000000727",
  ],
  Array [
    "Re render,1.6849999996338738,1.6849999996338738,1.6849999996338738,1.6849999996338738,1.6849999996338738",
  ],
  Array [
    "Hydrate,8.303500000329223,8.303500000329223,8.303500000329223,8.303500000329223,8.303500000329223",
  ],
  Array [
    "Complete render (mount + layout + paint),55.235999999968044,55.235999999968044,55.235999999968044,55.235999999968044,55.235999999968044",
  ],
  Array [
    "DOM element count,1475,1475,1475,1475,1475",
  ],
  Array [
    "DOM element count (no nested inline svg elements),706,706,706,706,706",
  ],
  Array [
    "React Fiber node count,3886,3886,3886,3886,3886",
  ],
  Array [],
  Array [
    "Type,fixtures",
  ],
  Array [
    "Render to string,11.901000000761996",
  ],
  Array [
    "Render to static markup (cannot be hydrated),10.560999999870546",
  ],
  Array [
    "String output size,184.64",
  ],
  Array [
    "String output size (gzip),41.16",
  ],
  Array [
    "Static markup output size,184.64",
  ],
  Array [
    "Static markup output size (gzip),41.16",
  ],
  Array [
    "Initial render,23.0890000000727",
  ],
  Array [
    "Re render,1.6849999996338738",
  ],
  Array [
    "Hydrate,8.303500000329223",
  ],
  Array [
    "Complete render (mount + layout + paint),55.235999999968044",
  ],
  Array [
    "DOM element count,1475",
  ],
  Array [
    "DOM element count (no nested inline svg elements),706",
  ],
  Array [
    "React Fiber node count,3886",
  ],
]
`;

exports[`cli should match snapshot 2`] = `
Array [
  Array [
    "Type | fixtures",
  ],
  Array [
    "===========================================",
  ],
  Array [
    "Render to string | 11.901000000761996",
  ],
  Array [
    "Render to static markup (cannot be hydrated) | 10.560999999870546",
  ],
  Array [
    "String output size | 184.64",
  ],
  Array [
    "String output size (gzip) | 41.16",
  ],
  Array [
    "Static markup output size | 184.64",
  ],
  Array [
    "Static markup output size (gzip) | 41.16",
  ],
  Array [
    "Initial render | 23.0890000000727",
  ],
  Array [
    "Re render | 1.6849999996338738",
  ],
  Array [
    "Hydrate | 8.303500000329223",
  ],
  Array [
    "Complete render (mount + layout + paint) | 55.235999999968044",
  ],
  Array [
    "DOM element count | 1475",
  ],
  Array [
    "DOM element count (no nested inline svg elements) | 706",
  ],
  Array [
    "React Fiber node count | 3886",
  ],
]
`;
